File: java/3 - lista encadeada/Lista.java
public class Lista {

  private Caixinha inicio;

  public Lista() {
    this.inicio = null;
  }

  public void inserir(int elemento) {
    Caixinha nova = new Caixinha();
    nova.setElemento(elemento);
    nova.setProximo(null);

    if (inicio == null) { // 1ª caixinha?
      inicio = nova;
    } else {
      Caixinha aux = inicio;
      while (aux.getProximo() != null) {
        aux = aux.getProximo();
      }
      aux.setProximo(nova);
    }
  }

  // esse metodo considera que a lista sempre tem elementos.
  public int retirar() {
    Caixinha aux = inicio;
    int elemento = aux.getElemento();
    inicio = aux.getProximo();
    return elemento;
  }

  public void listar() {
    if (inicio == null) {
      System.out.println("Nada para exibir. A lista está vazia!");
    } else {
      Caixinha aux = inicio;
      while (aux != null) {
        System.out.println("Elemento visto: " + aux.getElemento());
        aux = aux.getProximo();
      }
    }
  }

  public boolean isEmpty() {
    return (this.inicio == null);
  }
}

File: c/3 - lista encadeada/lista.c
#include "lista.h"
#include <stdlib.h>

void create(struct Lista *lista)
{
  lista->inicio = NULL;
}

void inserir(struct Lista *lista, int elemento)
{
  struct Caixinha *nova;
  struct Caixinha *aux;
  nova = (struct Caixinha *)malloc(sizeof(struct Caixinha));
  nova->proximo = NULL;
  nova->elemento = elemento;

  // esta vazia?
  if (lista->inicio == NULL)
  {
    lista->inicio = nova;
  }
  else
  {
    aux = lista->inicio;
    while (aux->proximo != NULL)
    {
      aux = aux->proximo;
    }
    aux->proximo = nova;
  }
}

int isEmpty(struct Lista lista)
{
  return (lista.inicio == NULL);
}

int retirar(struct Lista *lista)
{
  struct Caixinha *aux = lista->inicio;
  int elemento = aux->elemento;
  lista->inicio = aux->proximo;
  free(aux);
  aux = NULL;
  return elemento;
}